notebook/nbextensions.py

```diff
@@ -199,60 +192,444 @@ def install_nbextension(path, overwrite=False, symlink=False, user=False, prefix
+        else:
+            os.remove(full_dest)
+    
+    # Look through all of the config sections making sure that the nbextension
+    # doesn't exist.
+    config_dir = os.path.join(_get_config_dir(user=user, sys_prefix=sys_prefix), 'nbconfig')
+    cm = BaseJSONConfigManager(config_dir=config_dir)
+    for section in NBCONFIG_SECTIONS:
+        cm.update(section, {"load_extensions": {require: None}})
+
+
+def uninstall_nbextension_python(package,
+                        user=False, sys_prefix=False, prefix=None, nbextensions_dir=None,
+                        logger=None):
+    """Uninstall an nbextension bundled in a Python package.
+    
+    See parameters of `install_nbextension_python`
+    """
+    m, nbexts = _get_nbextension_metadata(package)
+    for nbext in nbexts:
+        dest = nbext['dest']
+        require = nbext['require']
+        if logger:
+            logger.info("Uninstalling {} {}".format(dest, require))
+        uninstall_nbextension(dest, require, user=user, sys_prefix=sys_prefix, 
+            prefix=prefix, nbextensions_dir=nbextensions_dir, logger=logger)
+
+
+def _set_nbextension_state(section, require, state,
+                           user=False, sys_prefix=False, logger=None):
+    """Set whether the section's frontend should require the named nbextension
+
+    Returns True if the final state is the one requested.
+
+    Parameters
+    ----------
+    section : string
+        The section of the server to change, one of NBCONFIG_SECTIONS
+    require : string
+        An importable AMD module inside the nbextensions static path
+    state : bool
+        The state in which to leave the extension
+    user : bool [default: False]
+        Whether to check the user's .jupyter/nbextensions directory.
+        Otherwise check a system-wide install (e.g. /usr/local/share/jupyter/nbextensions).
+    sys_prefix : bool [default: False]
+        Install into the sys.prefix, i.e. environment
+    logger : Jupyter logger [optional]
+        Logger instance to use
+    """
+    config_dir = os.path.join(
+        _get_config_dir(user=user, sys_prefix=sys_prefix), 'nbconfig')
+    cm = BaseJSONConfigManager(config_dir=config_dir)
+    if logger:
+        logger.info("{} {} extension {}...".format(
+            "Enabling" if state else "Disabling",
+            section,
+            require
+        ))
+    cm.update(section, {"load_extensions": {require: state}})
+    return cm.get(section).get(require) == state
+
+
+def _set_nbextension_state_python(state, package, user, sys_prefix,
+                                  logger=None):
+    """Enable or disable some nbextensions stored in a python package
+
+    Returns a list of whether the state was achieved (i.e. changed, or was
+    already right)
+
+    Parameters
+    ----------
+
+    state : Bool
+        Whether the extensions should be enabled
+    package : str
+        Importable Python package (no dotted-notation!) exposing the
```

Comment:
Agree with @minrk that the default should be the user level as it is in
4.1, rather than 4.2. I wasn't involved in the renaming of the method to `
_get_nbextension_metadata(package)` but if that is the case, I think it
makes sense.

On Wed, Mar 16, 2016 at 2:25 AM, Min RK <notifications@github.com> wrote:

> In notebook/nbextensions.py
> <https://github.com/jupyter/notebook/pull/879#discussion_r56302734>:
>
> > +
> > +
> > +def _set_nbextension_state_python(state, package, user, sys_prefix,
> > +                                  logger=None):
> > +    """Enable or disable some nbextensions stored in a python package
> > +
> > +    Returns a list of whether the state was achieved (i.e. changed, or was
> > +    already right)
> > +
> > +    Parameters
> > +    ----------
> > +
> > +    state : Bool
> > +        Whether the extensions should be enabled
> > +    package : str
> > +        Importable Python package (no dotted-notation!) exposing the
>
> And pedantry: what we're looking for here is a module, not a package.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/879/files/4db46915639d13f16ed3284793d072d2fe1201cd#r56302734>
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com